### DISCUSSION TITLE
Coordinates of SUMO nodes do not match configuration

### DISCUSSION THREAD

**Discussion by CiscoFrisco** [eclipse/mosaic#96] [Q&A]
Hi everyone,

So I'm currently working on a project using Eclipse MOSAIC and I'm building an application that requires me to map a RoadSideUnit to a network node.

My solution to this was to use the OperatingSystem interface to retrieve the GeoPosition of the RSU. Then, by using the getClosestNode() function from the RoutingModule and the retrieved position  I retrieve the closest node to the RSU. 

`String rsuNodeID = getOs().getRoutingModule().getClosestNode(getOs().getPosition()).getId();`

This code works but is giving me the wrong node. For some reason, the GeoPosition of the Nodes in the network is not correct. However, the GeoPosition of the vehicles traveling in the network is fine.

My SUMO scenario has no geographic coordinates since it is a custom scenario that I built myself. But MOSAIC is still able to somehow map the scenario with geographical coordinates. When converting the GeoPosition values from the vehicles to cartesian coordinates I get the expected values, but when doing so with the nodes GeoPositions, it gives me very big numbers that have nothing to do with the SUMO configuration.

For example, in the SUMO network configuration, I defined a node "n0" in the position (x:0,y:0).
The vehicles inserted are mapped to the geographic coordinates around (lat: 53,47, long:15,96) that when converted to cartesian coordinates give values very close to the position (x:0,y:0), as expected.
However, when using the following call to retrieve a node's position, 

`getOs().getRoutingModule().getNode("n0").getPosition().toCartesian()`

it returns geographic coordinates very close to (lat: 0,long: -1.5) which in return converts to the cartesian coordinates (x:-1924799.30,y:-5933566.87) which makes no sense when compared to the network configuration.

Since vehicles use the routes to spawn, and since routes are a combination of SUMO edges connecting to nodes, how can the vehicle's coordinates be correct and the node's coordinates not? Am I missing something?

I would really appreciate it if someone could help me with this issue.

P.S. Is there a way that by providing a node ID of an intersection node, I can retrieve the connecting lanes' positions? My goal with this mapping is ultimately for the RSU to know the position of each lane in the intersection. Does anyone know a way to do that with MOSAIC?

**Comment by kschrab** [comment]
MOSAIC's geo-projection converts latitude/longitude coordinates in the WGS84 system to Cartesian coordinates in the UTM system. During the import process, `scenario-convert` assumes that the coordinates of the SUMO network file are given in UTM. This results in those very weird coordinates as scenario-convert tries to format (x=0,y=0) from UTM to WGS84. To compensate this, the SUMO network file usually contains a `location.netOffset` tag. This offset is substracted from the coordinates resulting in valid UTM coordinates which can then be converted to valid WGS84 coordinates. However, this offset is usually set only if the network file was generated from real-word maps, such as OSM. 

During the simulation it works the same way: the plain x,y coordinates for the vehicles are returned from SUMO (e.g. x=0,y=0). Then, the `cartesianOffset` from the `scenario_config.json` is substracted from those coordinates, and transformed to WGS84 afterwards, resulting in "valid" coordinate such as (lat: 53,47, long:15,96). 

As a consequence, the `netoffset` of the network file needs to be equal to the `cartessianOffset` in the `scenario_config.json`. 

There might be two solutions now, but I think only on of those will work:
* Set the `cartessianOffset` in the `scenario_config.json` to `x=0, y=0`. However, I think this would break the WGS84<->UTM conversion and might result in errors or weird projection.
* Add/extend the `<location netOffset="x,y" ...>`  tag to the `net.xml` file of the SUMO scenario, and insert the same coordinates which are defined in the `cartessianOffset` in `scenario_config.json`. You should use UTM coordinates, such as `<location netOffset="-369461.94,-5798955.14" ...`.

If you still have troubles, please send us your scenario to mosaic@fokus.fraunhofer.de and we'll have a look on this.

**Reply by CiscoFrisco** [reply]
Are there any specific values that the netOffset parameter should receive or do they only need to be UTM coordinates that match in both the net.xml file and scenario_config.json? 

I tried the following location tag:

    <location netOffset="-563984.16,-5933566.87" convBoundary="-100.00,-100.00,100.00,100.00" origBoundary="0.00,0.00,0.00,0.00" projParameter="!"/>

Those netOffset values were already defined in the scenario_config.json. Don't remember where they came from, but they still don't solve the issue.

**Reply by CiscoFrisco** [reply]
One other quick question (not related to this): I noticed from [this](https://www.eclipse.org/mosaic/docs/simulators/application_mapping/#unit-identifiers) that MOSAIC uses some default identifiers for the objects used (e.g. rsu_0, veh_0, veh_1). Is there a way to use SUMO identifiers instead? I have vehicle definitions in a SUMO route file to define platooning vehicles, and their IDs give me information about the platoon. Using this ID as an object ID in MOSAIC is the only solution I can see.

**Reply by kschrab** [reply]
As you are using an artificial scenario, the exact offset is not that much of importance. They just have to match each other. You also would need to re-create the database using `scenario-convert --sumo2db -i path/to/net.xml --import-zone=33u`

**Reply by kschrab** [reply]
> One other quick question (not related to this): I noticed from [this](https://www.eclipse.org/mosaic/docs/simulators/application_mapping/#unit-identifiers) that MOSAIC uses some default identifiers for the objects used (e.g. rsu_0, veh_0, veh_1). Is there a way to use SUMO identifiers instead? 

Hm.. the feature of using pre-existing SUMO scenarios with MOSAIC is very new. We did not had the requirement yet to pass the original vehicle IDs from SUMO. Doing so would _at least_ brake the `IpResolver`, as it expects the IDs in this specific format. What you always could do is forking this repo and try to implement it by yourself. There you could replace the `TraciClient.VEHICLE_ID_TRANSFORMER` with an own implementation and see what fails then.

**Reply by CiscoFrisco** [reply]
I forgot to recreate the database. It worked now! Thank you very much!

**Reply by CiscoFrisco** [reply]
> > One other quick question (not related to this): I noticed from [this](https://www.eclipse.org/mosaic/docs/simulators/application_mapping/#unit-identifiers) that MOSAIC uses some default identifiers for the objects used (e.g. rsu_0, veh_0, veh_1). Is there a way to use SUMO identifiers instead?
> 
> Hm.. the feature of using pre-existing SUMO scenarios with MOSAIC is very new. We did not had the requirement yet to pass the original vehicle IDs from SUMO. Doing so would _at least_ brake the `IpResolver`, as it expects the IDs in this specific format. What you always could do is forking this repo and try to implement it by yourself. There you could replace the `TraciClient.VEHICLE_ID_TRANSFORMER` with an own implementation and see what fails then.

Is there a way of at least know the mapping of the MOSAIC IDs with SUMO IDs (e.g. veh_0 maps to SUMO ID: v0)? If possible that would actually be enough for me.

**Reply by kschrab** [reply]
> Is there a way of at least know the mapping of the MOSAIC IDs with SUMO IDs (e.g. veh_0 maps to SUMO ID: v0)? If possible that would actually be enough for me.

A quick-and-dirty solution would be to include `org.eclipse.mosaic:mosaic-sumo` into the `pom.xml` of your application and use 

```java
String sumoId = TraciClient.VEHICLE_ID_TRANSFORMER.toExternalId(getOs().getId())
```

**Reply by CiscoFrisco** [reply]
Thank you very much! It worked perfectly

**Comment by kschrab** [comment]
> P.S. Is there a way that by providing a node ID of an intersection node, I can retrieve the connecting lanes' positions? My goal with this mapping is ultimately for the RSU to know the position of each lane in the intersection. Does anyone know a way to do that with MOSAIC?

We store rather limited information about the network in our scenario-database, so I guess that would not be possible out-of-the-box. What you could do is to implement some pre-processing tool which extracts network information from the SUMO network file and stores it in a custom JSON file, which the RSU's application could load during the simulation.

**Reply by CiscoFrisco** [reply]
Thank you for your quick response! I hadn't thought of that yet. Will try to implement it!
